internal/restconfig/restconfig.go

```diff
@@ -26,58 +26,140 @@ import (
 	mcsv1a1 "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
 )
 
-type RestConfig struct {
+type restConfig struct {
 	Config      *rest.Config
 	ClusterName string
 }
 
+type configAndOverrides struct {
+	config    clientcmd.ClientConfig
+	overrides *clientcmd.ConfigOverrides
+}
+
 type Producer struct {
-	kubeConfig   string
-	kubeContext  string
-	kubeContexts []string
-	inCluster    bool
+	kubeConfig          string
+	kubeContext         string
+	kubeContexts        []string
+	defaultClientConfig *configAndOverrides
+	inCluster           bool
+	namespaceFlag       bool
 }
 
-func NewProducer() Producer {
-	return Producer{}
+func NewProducer() *Producer {
+	return &Producer{}
 }
 
-func NewProducerFrom(kubeConfig, kubeContext string) Producer {
-	return Producer{
+func NewProducerFrom(kubeConfig, kubeContext string) *Producer {
+	return &Producer{
 		kubeConfig:  kubeConfig,
 		kubeContext: kubeContext,
 	}
 }
 
-func (rcp *Producer) AddKubeConfigFlag(cmd *cobra.Command) {
-	cmd.PersistentFlags().StringVar(&rcp.kubeConfig, "kubeconfig", "", "absolute path(s) to the kubeconfig file(s)")
+func (rcp *Producer) WithNamespace() *Producer {
+	rcp.namespaceFlag = true
+
+	return rcp
 }
 
-// AddKubeContextFlag adds a "kubeconfig" flag and a single "kubecontext" flag that can be used once and only once.
-func (rcp *Producer) AddKubeContextFlag(cmd *cobra.Command) {
-	rcp.AddKubeConfigFlag(cmd)
-	cmd.PersistentFlags().StringVar(&rcp.kubeContext, "kubecontext", "", "kubeconfig context to use")
+func (rcp *Producer) SetupFlags(flags *pflag.FlagSet) {
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	loadingRules.DefaultClientConfig = &clientcmd.DefaultClientConfig
+
+	flags.StringVar(&loadingRules.ExplicitPath, "kubeconfig", "", "absolute path(s) to the kubeconfig file(s)")
+
+	// Default un-prefixed context
+	overrides := clientcmd.ConfigOverrides{ClusterDefaults: clientcmd.ClusterDefaults}
+	kflags := clientcmd.RecommendedConfigOverrideFlags("")
+	clientcmd.BindOverrideFlags(&overrides, flags, kflags)
+
+	// Support deprecated --kubecontext; TODO remove in 0.15
+	legacyFlags := clientcmd.RecommendedConfigOverrideFlags("kube")
+	legacyFlags.CurrentContext.BindStringFlag(flags, &overrides.CurrentContext)
+	_ = flags.MarkDeprecated("kubecontext", "use --context instead")
+
+	if !rcp.namespaceFlag {
+		_ = flags.MarkHidden("namespace")
+	}
+
+	rcp.defaultClientConfig = &configAndOverrides{
+		config:    clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, &overrides),
+		overrides: &overrides,
+	}
+}
+
+type PerContextFn func(clusterInfo *cluster.Info, namespace string) error
+
+// Runs the given function on the selected context.
+// The status reporter is only used for errors specific to this function;
+// the per-context function is expected to perform its own status reporting.
+func (rcp *Producer) RunOnSelectedContext(function PerContextFn, status reporter.Interface) error {
```

Comment:
Since this ends up tied to a status reporter anyway, I wonder if I shouldn’t just add `status` to `PerContextFn`.

Reply:
That seems reasonable.